scalar/struct_test.go

```diff
@@ -0,0 +1,63 @@
+package scalar
+
+import (
+	"strconv"
+	"testing"
+
+	"github.com/apache/arrow/go/v13/arrow"
+	"github.com/apache/arrow/go/v13/arrow/array"
+	"github.com/apache/arrow/go/v13/arrow/memory"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestStructMissingKeys(t *testing.T) {
+	tl := []struct {
+		dt          *arrow.StructType
+		input       any
+		expectPanic bool
+	}{
+		{dt: arrow.StructOf(arrow.Field{Name: "i64", Type: arrow.PrimitiveTypes.Int64}, arrow.Field{Name: "s", Type: arrow.BinaryTypes.String}), input: `{"i64": 1, "s": "foo"}`, expectPanic: false},
+		{dt: arrow.StructOf(arrow.Field{Name: "i64", Type: arrow.PrimitiveTypes.Int64}, arrow.Field{Name: "s", Type: arrow.BinaryTypes.String}), input: `{"i64": 1, "s": "foo", "extra":"bar"}`, expectPanic: true},
+	}
+
+	for idx, tc := range tl {
+		tc := tc
+		t.Run(strconv.Itoa(idx), func(t *testing.T) {
+			panicked := false
+			defer func() {
+				if t.Failed() {
+					return
+				}
+
+				if panicked && !tc.expectPanic {
+					t.Errorf("unexpected panic")
+				}
+				if !panicked && tc.expectPanic {
+					t.Errorf("expected panic")
+				}
+			}()
+
+			defer func() {
+				if r := recover(); r != nil {
```

Comment:
there's [`require.Panics`](https://pkg.go.dev/github.com/stretchr/testify/require#Panics) func family for such cases, FYI (with err can check err string, notpanics requires code not to panic, etc)